source/material_model/viscoelastic.cc

```diff
@@ -0,0 +1,700 @@
+            {
+              elastic_out->elastic_shear_moduli[i] = average_elastic_shear_modulus;
+            }
+
+        }
+
+      // Viscoelasticity section
+      if (in.current_cell.state() == IteratorState::valid && this->get_timestep_number() > 0 && in.strain_rate.size() > 0)
+        {
+          // Get old (previous time step) velocity gradients
+          std::vector<Point<dim> > quadrature_positions(in.position.size());
+          for (unsigned int i=0; i < in.position.size(); ++i)
+            quadrature_positions[i] = this->get_mapping().transform_real_to_unit_cell(in.current_cell, in.position[i]);
+
+          FEValues<dim> fe_values (this->get_mapping(),
+                                   this->get_fe(),
+                                   Quadrature<dim>(quadrature_positions),
+                                   update_gradients);
+
+          fe_values.reinit (in.current_cell);
+          std::vector<Tensor<2,dim> > old_velocity_gradients (Quadrature<dim>(quadrature_positions).size(), Tensor<2,dim>());
+          fe_values[this->introspection().extractors.velocities].get_function_gradients (this->get_old_solution(),
+                                                                                         old_velocity_gradients);
+
+          MaterialModel::AdditionalMaterialOutputsStokesRHS<dim>
+          *force_out = out.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >();
+
+          for (unsigned int i=0; i < in.position.size(); ++i)
+            {
+              // Get old stresses from compositional fields
+              SymmetricTensor<2,dim> stress_old;
+              for (unsigned int j=0; j < SymmetricTensor<2,dim>::n_independent_components; ++j)
+                stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)] = in.composition[i][j];
+
+              // Calculate the rotated stresses
+              // Rotation (vorticity) tensor (equation 25 in Moresi et al., 2003, J. Comp. Phys.)
+              const Tensor<2,dim> rotation = 0.5 * ( old_velocity_gradients[i] - transpose(old_velocity_gradients[i]) );
+
+              // Recalculate average values of viscosity, elastic shear modulus and viscoelastic (effective) viscosity
+              const std::vector<double> composition = in.composition[i];
+
+              // Average viscosity
+              const double average_viscosity = calculate_average_viscosity(composition,
+                                                                           viscosities,
+                                                                           viscosity_averaging);
+              // Average elastic shear modulus
+              const double average_elastic_shear_modulus =
+                calculate_average_elastic_shear_modulus(composition,
+                                                        elastic_shear_moduli,
+                                                        viscosity_averaging);
```

Comment:
You can use the one from `elastic_out->elastic_shear_moduli[i]`

Reply:
@jdannberg - For some reason this broke the tests. I am going to leave the current structure for now, but happy to revisit this later.